File: code_python/PPCA.py
import numpy as np

def PPCA(t, q, epsilon = 0.001, method = 'ML'):
    d, N = t.shape
    Mu = np.mean(t, axis = 1).reshape(d, 1)
    S = np.zeros((d, d))

    for i in range(N):
        S += np.dot((t[:, i].reshape(d, 1) - Mu), (t[:, i].reshape(d, 1) - Mu).T)
    S /= N

    if method == 'ML':
        value, U = np.linalg.eig(S)
        U = U[:, 0: q]
        sigma_square = np.sum(value[q: d]) / (d - q)
        Lambda = np.diag(value[0: q])
        W = np.dot(U, np.sqrt(Lambda - sigma_square * np.identity(q)))
        W = W.real

        M = np.dot(W.T, W) + sigma_square * np.identity(q)
        X = np.dot(np.dot(np.linalg.inv(M), W.T), (t - Mu))

    if method == 'EM':
        value, U = np.linalg.eig(S)
        U = U[:, 0: q]
        U = U.real
        Lambda = np.diag(value[0: q])

        W_new = np.dot(U, np.sqrt(Lambda))
        W_old= np.dot(U, np.sqrt(Lambda))
        W_new = W_new.real
        W_old = W_old.real
        sigma_square_new = 1
        sigma_square_old = 5

        while (np.sqrt(np.sum(np.square(W_new - W_old))) > epsilon) | (abs(sigma_square_new-sigma_square_old) > epsilon):
            M = np.dot(W_new.T, W_new) + sigma_square_new * np.identity(q)
            W_old = W_new
            sigma_square_old = sigma_square_new
            W_new = np.dot(np.dot(S, W_old), np.linalg.inv(sigma_square_old * np.identity(q) + np.dot(np.linalg.inv(M), np.dot(W_old.T, np.dot(S, W_old)))))
            sigma_square_new = np.trace(S - np.dot(S, np.dot(W_old, np.dot(np.linalg.inv(M), W_new.T))))/d

            C = np.dot(W_new, W_new.T) + sigma_square_new * np.identity(d)
            L = -N * (d * np.log(np.pi) + np.log(np.linalg.det(C)) + np.trace(np.dot(np.linalg.inv(C), S))) / 2
            print('updated log-likelihood: ', L)

        M = np.dot(W_new.T, W_new) + sigma_square_new * np.identity(q)
        X = np.dot(np.linalg.inv(M), np.dot(W_new.T, (t - Mu)))

    return X

File: code_R/PPCA.R
PPCA = function(t, q, epsilon=0.001, method = 'ML'){
  
#Input data:t
  t = as.matrix(t)
  
#t has the shape of d*N 
  d = dim(t)[1]
  N = dim(t)[2]
  
#Mu is the average of each rows
  Mu = rowSums(t) / N
  
#Compute the variance S
  S = matrix(data = 0, nrow = d, ncol = d)
  for (i in 1 : N){S = S + (t[, i] - Mu) %*% t(t[, i] - Mu)}
  S = S / N
  
#1.Use the result of the maximum likelihood directly
  if (method == 'ML'){
 
#Calculate the eigenvalue and eigenvector of S
  ev = eigen(S)
  
#Calculate sigma square
  if (q + 1 == d){sigma = sum(ev$values[d]) / (d - q)}
  if (q + 1 < d){sigma = sum(ev$values[(q + 1) : d]) / (d - q)}
  print(sigma)
  
#Calculate W
  U = ev$vectors[, 1 : q]
  if (q == 1){lambda = ev$values[1]}
  if (q > 1){lambda = diag(ev$values[1 : q])}
  W = U %*% sqrt(lambda - sigma * diag(q))
  print(W)
  
#Calculate the desired data X
  M = t(W) %*% W + sigma * diag(q)
  X = solve(M) %*% t(W) %*% (t-Mu)
  }

#2.Use EM algorithm
  if (method == 'EM'){
  
#Generate inital W and sigma_square 
   ev = eigen(S)
   U = ev$vectors[, 1 : q]
   if (q == 1){lambda = ev$values[1]}
   if (q > 1){lambda = diag(ev$values[1 : q])}
   
   W_old = U %*% sqrt(lambda)
   W_new = U %*% sqrt(lambda)
   sigma_old = 2
   sigma_new = 5
  
#EM algorithm
  while (sqrt(sum((W_old - W_new) ^ 2)) > epsilon || abs(sigma_new - sigma_old) > epsilon){
    M = t(W_new) %*% W_new + sigma_new * diag(q)
    W_old = W_new
    sigma_old = sigma_new
    W_new = S %*% W_old %*% solve(sigma_old * diag(q) + solve(M) %*% t(W_old) %*% S %*% W_old)
    sigma_new = sum(diag(S - S %*% W_old %*% solve(M) %*% t(W_new)))/d
    
    print(W_new)
    print(sigma_new)
  }
  
#Calculate the desired data X
  M = t(W_new) %*% W_new + sigma_new * diag(q)
  X = solve(M) %*% t(W_new) %*% (t - Mu)
  }
  
  return(X)
}






